source/lac/lapack_full_matrix.cc

```diff
@@ -172,6 +178,8 @@ LAPACKFullMatrix<number>::vmult (
   const Vector<number> &v,
   const bool            adding) const
 {
+  Threads::Mutex::ScopedLock lock (mutex);
```

Comment:
nice job making these functions thread-safe!

Reply:
Though on second thought, I guess you do this because you are using the `work` array, right? You only need `work` in two of the three `case` blocks below, so maybe the right place to lock would be those branches where you actually use it? The regular `gemv` calls in `case matrix:` and `case inverse_matrix:` are *already* thread-safe because they can run on multiple threads at the same time, no locking needed.

Reply:
correct, only for using `mutable`, it was just easier to add it at the top.

Reply:
I think it would be worthwhile moving it, though, to allow for the parallelism possible when just using a regular mat-vec.